src/main/java/net/neoforged/neoforge/transfer/resources/ItemResource.java

```diff
@@ -0,0 +1,331 @@
+    }
+
+    /**
+     * A helper method to quickly construct an {@link ItemStack} from a ResourceStack
+     *
+     * @param resourceStack The resource stack with the item resource and amount
+     * @return A new item stack with the same size as the resourceStack.
+     */
+    public static ItemStack itemStackOf(ResourceStack<ItemResource> resourceStack) {
+        return resourceStack.resource().toStack(resourceStack.amount());
+    }
+
+    /**
+     * This is used only for registry, you should not use this method!
+     */
+    @ApiStatus.Internal
+    public static ItemResource createDefaultInstance(ItemLike item) {
+        if (item.asItem() == Items.AIR) return EMPTY;
+        return new ItemResource(new ItemStack(item));
+    }
+
+    /**
+     * Creates an ItemResource using the default or copy of the passed in item stack.
+     * 
+     * @param itemStack stack to copy
+     * @return If there were no patches on the stack's data components, the item's default resource will be returned, otherwise a new instance with the copied stack.
+     */
+    public static ItemResource of(ItemStack itemStack) {
+        if (itemStack.isEmpty()) return EMPTY;
+
+        if (itemStack.isComponentsPatchEmpty())
+            return itemStack.getItem().getDefaultResource();
+
+        return new ItemResource(itemStack.copyWithCount(1));
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your item is registered
+     * 
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    public static ItemResource of(ItemLike item) {
+        if (item.asItem() == Items.AIR) return EMPTY;
+        return item.asItem().getDefaultResource();
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your item is registered
+     * 
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    public static ItemResource of(Holder<Item> item, DataComponentPatch patch) {
+        if (item.value() == Items.AIR) return EMPTY;
+        if (patch.isEmpty()) return item.value().getDefaultResource();
+        return item.value().getDefaultResource().withPatch(patch);
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your item is registered
+     * 
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    public static ItemResource of(Holder<Item> item) {
+        return of(item.value());
+    }
+
+    /**
+     * A wrapped {@link ItemStack} which must never be modified.
+     */
+    private final ItemStack innerStack;
```

Comment:
I'm not sure this should be a wrapper over an ItemStack, versus having the Item and PatchedDataComponentMap directly.

Reply:
That's fair. I was looking at how Fabric implemented their variants as a reference while noodling over some details in the fluid rework and noticed they were using the item and data component map directly; that felt more reasonable to me but I can understand the practicality of doing it this way.

Another reason I brought this up is that if we're introducing a ResourceStack here to represent amounts of resources, then it might be worth reconsidering having a FluidStack in the fluid rework. This would make FluidResource necessarily be different, so I also commented here in order to determine whether consistency or convenience/practicality was more important.